### ISSUE TITLE
On highlight/select text - cannot unhighlight/deselect text on click - need to do ESC

### ISSUE THREAD

**Issue by redapplewithleaves** [Microsoft/vscode#28636]
- VSCode Version: Code 1.13.0 (376c52b955428d205459bea6619fc161fc8faacf, 2017-06-08T16:43:13.058Z)
- OS Version: Windows_NT ia32 6.1.7601
- Extensions: 

|Extension|Author|Version|
|---|---|---|
|git-easy|bibhasdn|1.9.1|
|file-icons|file-icons|1.0.3|
|auto-close-tag|formulahendry|0.4.2|
|auto-rename-tag|formulahendry|0.0.12|
|minify|HookyQR|0.3.0|
|Angular2|johnpapa|2.3.2|
|vscode-github|KnisterPeter|0.16.2|
|vs-deploy|mkloubert|9.25.0|
|prettify-json|mohsen1|0.0.3|
|debugger-for-chrome|msjsdiag|3.1.3|
|vscode-run-git-difftool|narekmal|0.0.4|
|gitblame|waderyan|1.11.2|
|bootstrap-3-snippets|wcwhitehead|0.0.9|;

---

Steps to Reproduce:

1. Open a file
2. Click Ctrl+A 
3. Click anywhere - highlight still stays and can't do anything else

**Comment by ramya-rao-a** [comment]
@redapplewithleaves Do you any kind of codelens on your file?

I can repro this only when clicking on the line occupied by the codelens
Clicking anywhere else removes the selection.

**Comment by redapplewithleaves** [comment]
I haven't used code lens. Not quite sure its on...

Here is a screencast -> Each time the cursor change, I'm clicking on the the mouse. It takes about 7 or 8 times to clear the select. 
![2017-06-13](https://user-images.githubusercontent.com/9419314/27102023-a765127e-5038-11e7-8ebf-44978320dc5a.gif)

**Comment by ramya-rao-a** [comment]
@redapplewithleaves Can you share your settings?

**Comment by rebornix** [comment]
@redapplewithleaves could you please try disable drag and drop in the editor by setting `editor.dragAndDrop` to `false` and see if it happens still? Our mouse target detecting seems not smart enough.

**Comment by redapplewithleaves** [comment]
@rebornix   `editor.dragAndDrop`: `false` did the trick. 

Thank you so much!

This is my overwrite for default settings:

> // Place your settings in this file to overwrite the default settings
> {
>     "typescript.tsdk": "node_modules/typescript/lib",
>     "workbench.iconTheme": "file-icons",
>     "window.zoomLevel": -1,
>     "editor.dragAndDrop": false
> }

**Comment by russianryebread** [comment]
This appears to be happening on Mac as well:
- VSCode Version: Code 1.13.0 (376c52b, 2017-06-08T16:34:53.678Z)
- OS Version: Darwin x64 16.6.0

**Comment by johneriksson** [comment]
Same problem

- VSCode 1.13.1
- MacOS Sierra 10.12.5

`editor.dragAndDrop: false` fixed it for me too

**Comment by joostmeulenbeld** [comment]
Same here, same solution on Linux (ubuntu 16.04)
--
- VSCode Version: Code 1.13.1 (379d2efb5539b09112c793d3d9a413017d736f89, 2017-06-14T18:18:03.540Z)
- OS Version: Linux x64 4.4.0-83-generic
- Extensions: 

|Extension|Author|Version|
|---|---|---|
|matlab|Gimly81|0.5.0|
|latex-workshop|James-Yu|2.9.1|
|python|donjayamanne|0.6.7|
|code-spell-checker|streetsidesoftware|1.2.1|;

**Comment by r3wt** [comment]
i am having same issue on multiple OS!!!!

- Windows 10
- Windows 7
- Ubuntu 14.04
- Ubuntu 16.04 with Budgie Desktop (Ubuntu budgie-remix)

**Comment by tylerb** [comment]
Same issue. Same solution.

**Comment by rebornix** [comment]
Hi all, sorry for coming back to it too late, and thank you for your patience. I've tried hard to reproduce this issue but failed. I tested with latest Stable/Insider on three different platforms

1.18.1 on Ubuntu 16.04
![linux-dnd](https://user-images.githubusercontent.com/876920/33279662-37667c48-d354-11e7-8e51-ce16976c2aae.gif)

1.19-insiders on macOS High Sierra
![mac-dnd](https://user-images.githubusercontent.com/876920/33279679-4f5586d2-d354-11e7-813d-d63e5ed64b98.gif)

1.18.1 on Windows 10
![win-dnd](https://user-images.githubusercontent.com/876920/33279876-0e678516-d355-11e7-8beb-9f53ef58ddde.gif)

I tried to drag and drop then click, select all then click, drag and hold then click, click on selection, click on end of line (which is not on selection but on selected lines), etc but can't reproduce. I have a feeling it might be a problem with Chromium where the drag/drop/click event sequence is not guaranteed but it's just my wild guess. Can anyone help me with troubleshooting by

* Update to latest version (I suppose most of you already did)
* Some gif like @redapplewithleaves did
* Provide specific platform/version info
* Launch without extensions
* Not too complicated user setting

I may ask for too much but appreciate any help from you.

**Comment by russianryebread** [comment]
As far as I can tell, it's no longer an issue in the latest stable version: `1.18.1` (MacOS)

**Comment by r3wt** [comment]
in version `1.18.1` (Windows 10)

![screen1](https://user-images.githubusercontent.com/4483325/33284479-dcfc352e-d374-11e7-8435-fa3474509b8e.gif)
Screen 1:
1. select, drag and drop, right click, then click. takes two clicks to deselect if right click occurs first.

![screen2](https://user-images.githubusercontent.com/4483325/33284480-dd73800c-d374-11e7-886c-94b18066cc0d.gif)
2. select, drag and drop, click. deselect as expected without much lag(granted i am on extremely highend ~$2000 laptop)

you can see in both screen caps, when drag and drop css with a color box, the color box is left behind. very ugly.

**Comment by ctusch** [comment]
I'm using Ubuntu 16.04.3 LTS 64-bit and VSCode 1.18.1. I've removed all my user settings and started VSCode with --disable-extensions. The bug still occurs:
![peek 2017-11-27 21-17](https://user-images.githubusercontent.com/6949649/33287299-90c6a772-d3b8-11e7-8db1-8c7c28bf27b9.gif)
First I try single clicks on selected lines. Then I click on an unselected line which deselects the text. The 2nd time I do a fast double click on a selected line which also deselects the text successfully.

I've added event listeners too but could witness anything unusual except that a normal speed double click triggers the dblclick event but doesn't trigger a deselect. To deselect I have to double click even faster. Seems like it's handled in VSCode.

**Comment by alexrussell** [comment]
It would be useful to decrease drag sensitivity (or increase distance moved before a click is detected as a drag, however you wish to phrase it). Right now, if I click inside a selection with a 0px move, it's a click and deselects the selection. However, if I move horizontally or vertically by even 1 pixel (which it would appear my standard click on my magic mouse tends to do) it's a drag so no deselection happens.

**Comment by alexrussell** [comment]
In fact I have a feeling my mouse isn't even moving by 1 pixel (though I still think changing the drag distance will be useful for people), I have a gif here of me selecting and click within the selection (and to the right of but 'within' the selection) and I can't see the cursor move. However, every time it doesn't deselect, the cursor turns from a text caret to a normal arrow (which is indicative of a drag, among other thing admittedly), so I still think that somehow it's going into "drag mode", albeit not necessarily due to the cursor having moved.

![drag](https://user-images.githubusercontent.com/412784/36844594-4a742948-1d4b-11e8-9ada-58daeef0746c.gif)

**Comment by r3wt** [comment]
💯 i agree that would be useful. this is an extremely annoying problem

On Mar 1, 2018 6:16 AM, Alex Russell <notifications@github.com> wrote:

It would be useful to decrease drag sensitivity (or increase distance moved before a click is detected as a drag, however you wish to phrase it). Right now, if I click inside a selection with a 0px move, it's a click and deselects the selection. However, if I move horizontally or vertically by even 1 pixel (which it would appear my standard click on my magic mouse tends to do) it's a drag so no deselection happens.

**Comment by timdmackey** [comment]
alexrussell is correct, the issue here is with drag sensitivity. I just tested this and if I am very careful to not move the mouse, deselect works. If I work normally though, my mouse usually moves a tiny bit and so drag and drop is activated rather than the desired result of deselecting.

**Comment by r3wt** [comment]
bug still occuring for me on latest build as well. i thought it had been fixed, but i was mistaken. they just got a little bit better at making it not happen, but it still happens.

**Comment by alexrussell** [comment]
@timdmackey Yeah I think specifically it's the use of a magic mouse. Because the whole mouse clicks when you click, it appears to be easier to accidentally "move" it in the process (especially if you have your general mouse moving sensitivity turned up like I do).

I see similar issues with Firefox (click on a `<select>` element and a fair amount of the time it "selects" the value and closes straight away, because I moved 1 pixel and am now in drag-to-select-an-option mode, or right click to bring up context menu and accidentally go back in the page for the same reason). I don't use Sublime but I also see people in the office have issues with deselecting a selection (this exact issue), and it's quite possible it's the same thing - any amount of movement = drag.

But yeah, the magic mouse/drag thing is really just anecdotal. I can't say with certainty this is the culprit.

**Comment by timdmackey** [comment]
@alexrussell I also use a Magic Mouse, so I tried using a regular logitech mouse and was able to replicate the issue with that mouse too. If I'm working really quickly I tend to move my mouse a bit sometimes when I click, which causes the drag and drop to activate.

I think the drag and drop activation just needs to have a couple pixels of movement allowance before it activates.

I ended up disabling drag and drop in my settings because I don't like using the feature anyways.

**Comment by alexrussell** [comment]
@timdmackey yeah I guess it's not *just* a Magic Mouse that'll have the issue, but is certainly seems worse on a Magic Mouse than a touchpad for example. but, as I say, it's pretty anecdotal and definitely not mouse specific.

**Comment by robotboy655** [comment]
Would appreciate this getting fixed, sacrificing drag'n'drop to fix this is not that great of a solution.

This happens even with 1 line/block being fully selected, not just when the entire file is, which makes it that much more annoying.

**Comment by DominicTobias** [comment]
Why would you not want to deselect something when clicking inside the selection? This seems nothing to do with a bug but by some bizarre UX decision? It still exists in 1.25.

**Comment by garrettw** [comment]
I personally have only had this issue with a Magic Mouse so far. I'll update if it happens with my Razer mouse.

**Comment by DominicTobias** [comment]
Happens to me on a Windows 7 machine with some generic windows mouse, almost never unselects when clicking in a selected area really annoying.

**Comment by kyleloyka** [comment]
To echo what @garrettw  said, on MacOS I've seen this behavior with the Magic Mouse and Track Pad, but not with the Mighty Mouse.

**Comment by kyleloyka** [comment]
setting `"editor.dragAndDrop": false` "fixed" the issue for me. I agree with what was said above about decreasing the sensitivity of editor drag and drop as a more permanent fix. It seems the mouse moving ever-so-slightly causes the drag and drop functionality to be triggered.

I was seeing this behavior on MacOS with my Magic Mouse and built in trackpad. I was not seeing this behavior with my Mighty Mouse.

**Comment by zhuhang-jasper** [comment]
I don't think it's because of sensitivity of drag n drop. 
Even when the mouse is not moving AT ALL, the text still doesn't get deselected.

For now, what I do is: (without using `"editor.dragAndDrop": false`)
- press Esc key to deselect
- click at lines where text not selected (except case when whole document is highlighted)

**Comment by alexrussell** [comment]
Yes but are you 100% the cursor doesn't move even one pixel? Because this happens for me when using the magic mouse a lot more than the trackpad (I haven't tried a "traditional" mouse as I don't have one) I feel that it's that my cursor is moving an undetectable (by eye) amount when the click happens, which makes Code think it's a drag.

I imagine setting drag sensitivity to even 2 pixels would fix this. But I unfortunately can't prove this.

**Comment by rebornix** [comment]
The `drag` event was sent out when the mouse moves without releasing, but we didn't throttle or had any sensitivity adjustment. Sounds like that's what we should do, especially for Magic Mouse or any pointer device which is more likely to move a little bit when we click (it's hard to click without any movement at all by using fingers IMO).

One thing I'm not sure about if that would lead to any problem once we decrease the sensitivity, like if users fail to drag and drop anymore (for example, drag a single character which is probably single digit pixels.)

**Comment by garrettw** [comment]
I wonder if waiting for just 2 or 3 pixels of movement before starting the drag action would be sufficient.

**Comment by danieldbird** [comment]
Or delaying the drag and drop event only if the mouse has been held down > 100ms?

**Comment by armchairdeity** [comment]
I'm glad to now I'm not crazy...

Just to add my $0.02 to the conversation, if you set the trackpad to tap-to-click and put the mouse pointer over the middle of a block of selected text and tap... it *always* delects... I have a Magic Mouse and it pretty much _never_ works, even if I move my hand off the mouse and just activate the click with my fingertip. I can 99% guarantee you that the mouse itself did not move. I wonder if the mechanical click in the microswitch is enough to jostle the laser enough to give a false motion signal?

**Comment by alexrussell** [comment]
This is 100% not a solution to the issue, but as a suggestion to work out whether it's what you say, try turning the mouse upside down so the laser isn't doing its job and then click with your nail (so the swipe-scrolling mechanism has a minimal change of kicking in).

When I do that, it deselects 100% of the time, so I maintain that the magic mouse is just prone to a 1px (and apparently visually undetectable) movement during a click much of the time as we're all starting to agree.

And just going back to my earlier point, I know it's not exactly Code's place to fix the issues with certain mice, but I can't see not detecting a drag until the mouse has moved at least 2 pixels as being a thing that'd break anything.

**Comment by danieldbird** [comment]
FWIW I'm using a Logitech 403 mouse. I can say with almost 100% certainty it is the inadvertant moving of the mouse at least 1px whilst clicking that causes this issue.

**Comment by keithcelt** [comment]
I use an external trackpad on OS X and use Emacs navigation in the editor pane. Until the latest update the `editor.dragAndDrop` fix worked for me. Now the problem is back. Whenever I use keyboard navigation the cursor selects every character I navigate past. Obviously this isn't really a dragging issue since the trackpad doesn't move during keyboard navigation.

**Comment by quyenvsp** [comment]
This error still happening on 2022 Windows 11  22000.556 with latest VScode version 1.65.2, on same computer I tried install VSCode on HyperV machine and all working fine.
So I think because of old setting, tried Complete Uninstall, delete %AppData%\Code, .vscode,...restart computer then install again. VScode now complete new install but still not working, and dont know what is reason.

So final I go with `setting editor.dragAndDrop to false` and forgot about it. Hope it save time of someone want to try re-install solution.

**Comment by WrecksMedia** [comment]
Experienced the same issue, disabling Drag and Drop under the Text Editor settings resolved the problem.
